src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export enum PrincipalType {
+  Users = 'users',
+  Groups = 'groups',
+}
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export type Permissions = Record<string, Principals>;
+
+export interface TransformedPermission {
+  type: string;
+  name: string;
+  permissions: string[];
+}
+
+const addToPrincipals = ({
+  principals = {},
+  users,
+  groups,
+}: {
+  principals: Principals;
+  users?: string[];
+  groups?: string[];
+}) => {
+  if (!!users) {
+    if (!principals.users) {
+      principals.users = [];
+    }
+    principals.users = Array.from(new Set([...principals.users, ...users]));
+  }
+  if (!!groups) {
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deleteFromPrincipals = ({
+  principals,
+  users,
+  groups,
+}: {
+  principals?: Principals;
+  users?: string[];
+  groups?: string[];
+}) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+const checkPermission = (currentPrincipals: Principals | undefined, principals: Principals) => {
```

Comment:
The naming of the function and its parameters are not intuitive, consider changing `currentPrincipals` to `allowedPrincipals`, and `principals` to `requesterPrincipal` 

btw, `principals` indicates it is a collection, while it is indeed one single identity. I think the use case is only check if a request identity has permission to have access to a saved object, so maybe not using principal type for requester?

Reply:
Changing the name of the variables makes sense, thanks! For the principal type, I think its purpose is to make it consistent with the UI, the UI need to display the principal type: user or group, so when adding/removing/checking permissions, the caller need to use principal type explicitly.